### PR TITLE
use explicitly non-returning GPU atomics

### PR DESCRIPTION
**Pull request by jeffdaily**
Enables an important performance optimization for ROCm, in light of the discussion in #41028.

CC @jithunnair-amd @sunway513

### CODE REVIEW

File: aten/src/ATen/native/cuda/Sorting.cu [hunk 1]
```diff
@@ -136,7 +136,7 @@ __global__ void gatherMedian(
   }
   __syncthreads();
   if (nan_count > 0) {
-    atomicAdd(&num_nan, nan_count);
+    atomicAddNoReturn(&num_nan, nan_count);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Sorting.cu:139]: I think atomicAdd is here by mistake, and you can replace it by `gpuAtomicAddNoReturn`

File: aten/src/THC/THCAtomics.cuh [hunk 2]
```diff
@@ -281,6 +281,45 @@ static inline __device__ void atomicAdd(bool *address, bool val) {
...
+static inline __device__ void gpuAtomicAddNoReturn(int8_t *address, int8_t val) { gpuAtomicAdd(address, val); }
+static inline __device__ void gpuAtomicAddNoReturn(int16_t *address, int16_t val) { gpuAtomicAdd(address, val); }
+static inline __device__ void gpuAtomicAddNoReturn(int32_t *address, int32_t val) { gpuAtomicAdd(address, val); }
+static inline __device__ void gpuAtomicAddNoReturn(int64_t *address, int64_t val) { gpuAtomicAdd(address, val); }
+static inline __device__ void gpuAtomicAddNoReturn(bool *address, bool val) { gpuAtomicAdd(address, val); }
+static inline __device__ void gpuAtomicAddNoReturn(at::Half *address, at::Half val) { gpuAtomicAdd(address, val); }
+static inline __device__ void gpuAtomicAddNoReturn(at::BFloat16 *address, at::BFloat16 val) { gpuAtomicAdd(address, val); }
+static inline __device__ void gpuAtomicAddNoReturn(double *address, double val) { gpuAtomicAdd(address, val); }
+template<typename T>
+static inline __device__ void atomicAddNoReturn(c10::complex<T> *address, c10::complex<T> val) { gpuAtomicAdd(address, val); }
```
Comment (ngimel) [inline aten/src/THC/THCAtomics.cuh:303]: if you get rid of atomicAdd in Sorting.cu (replace it with gpuAtomicAddNoReturn) you won't need these functions here.

**Comment by ngimel** [comment]
Importing to see how internal builds go.

File: aten/src/THC/THCAtomics.cuh (hunk 2 above)
Comment (jeffdaily) [reply aten/src/THC/THCAtomics.cuh]: Applied suggestion in de042d3a2f829520ae0dfa5b6874dd3a68aabe15.

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**